static/index.js

```diff
@@ -0,0 +1,294 @@
+document.getElementById("botSelected").onchange = function (event) {
+  console.log(getCurrentBot());
+};
+
+document.addEventListener(
+  "keydown",
+  (event) => {
+    var key = "";
+    const keyName = event.code;
+    switch (keyName) {
+      case "KeyW":
+        keys.w = true;
+        key = "forward";
+        break;
+      case "KeyA":
+        keys.a = true;
+        key = "right";
+        break;
+      case "KeyS":
+        keys.s = true;
+        key = "back";
+        break;
+      case "KeyD":
+        keys.d = true;
+        key = "left";
+        break;
+      case "Space":
+        keys.space = true;
+        key = "jump";
+        break;
+    }
+    if (key != "") {
+      keylogger(keys);
+      send({
+        action: "move",
+        botId: getCurrentBot(),
+        data: { operation: key, state: true },
+      });
+    }
+  },
+  false
+);
+
+document.addEventListener(
+  "keyup",
+  (event) => {
+    if (event.repeat) {
+      return;
+    }
+    const keyName = event.code;
+    switch (keyName) {
+      case "KeyW":
+        keys.w = false;
+        key = "forward";
+        break;
+      case "KeyA":
+        keys.a = false;
+        key = "right";
+        break;
+      case "KeyS":
+        keys.s = false;
+        key = "back";
+        break;
+      case "KeyD":
+        keys.d = false;
+        key = "left";
+        break;
+      case "Space":
+        keys.space = false;
+        key = "jump";
+        break;
+    }
+    if (key != "") {
+      keylogger(keys);
+      send({
+        action: "move",
+        botId: getCurrentBot(),
+        data: { operation: key, state: false },
+      });
+    }
+  },
+  false
+);
+
+// looks like a function but is actually just to trigger the listeners after the server is online
+function listenSocket() {
+  socketserver.onopen = function (event) {
+    console.log("connected");
+    connected = true;
+    serverOnline(true);
+  };
+  socketserver.onmessage = function (event) {
+    console.log(event.data);
+    var message = JSON.parse(event.data);
+    switch (message.action) {
+      case "coords":
+        setCoords(message);
+        break;
+      case "health":
+        setHealth(message);
+        break;
+      case "hunger":
+        setHunger(message);
+        break;
+      case "started":
+        botOnline(message.action);
+        break;
+    }
+  };
+  socketserver.onclose = function (event) {
+    console.log("connection closed");
+    serverOnline(false);
+  };
+}
+
+// functions
+
+function say(message) {
+  send({
+    action: "say",
+    botId: getCurrentBot(),
+    message: message,
+  });
+}
+function kill() {
+  send({
+    action: "kill",
+    botId: getCurrentBot(),
+  });
+}
+
+function place() {
+  send({
+    action: "place",
+    botId: getCurrentBot(),
+  });
+}
+function attack() {
+  send({
+    action: "attack",
+    botId: getCurrentBot(),
+  });
+}
+function setServer(address, port) {
+  var address = document.getElementById("server").value;
```

Comment:
This works, but it feels a little...weird? to me - can we pull this and the port from env vars instead?

Reply:
this is on the website itself. it grabs the data in the form elements then sends them to the node server to use as alternates. the point of the entire panel was more for testing because of how minecraft randomizes port when opening a game to lan. the entire panel could probably be turned into a modal/popup